# Student Performance — Exploratory Data Analysis

## Purpose

This notebook explores the cleaned Student Performance Factors dataset
to investigate how study habits, attendance, and other factors relate
to exam scores.

The analysis will include:
- Reviewing distinct values and the `Unknown` category used during cleaning.
- Exploring summary statistics and distributions.
- Creating visualisations to address the project’s research questions.
- Recording findings and limitations.

The cleaned dataset is loaded from:
`data/processed/StudentPerformanceFactors_cleaned.csv`

Relationships observed in this analysis do not prove cause and effect.


## Import Libraries

Import the libraries needed to load, explore and visualise the cleaned dataset.

In [1]:
# Data handling and numerical operations
import pandas as pd
import numpy as np

# Data visualisation
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

# File paths
from pathlib import Path

## Check the Working Directory

Check the notebook's current working directory before setting the path
to the cleaned dataset.

In [2]:
# Display the current working directory
Path.cwd()

WindowsPath('c:/projects/student-performance-analysis/jupyter_notebooks')

## Load the Cleaned Dataset

Load the cleaned CSV from the project's `data/processed` folder
and preview the first five rows.

In [3]:
# Locate the cleaned dataset from the notebook's working directory
data_path = Path("..") / "data" / "processed" / "StudentPerformanceFactors_cleaned.csv"

# Load the cleaned dataset
df = pd.read_csv(data_path)

# Preview the first five rows
df.head()

,Hours_Studied,Attendance,Parental_Involvement,Access_to_Resources,Extracurricular_Activities,Sleep_Hours,Previous_Scores,Motivation_Level,Internet_Access,Tutoring_Sessions,Family_Income,Teacher_Quality,School_Type,Peer_Influence,Physical_Activity,Learning_Disabilities,Parental_Education_Level,Distance_from_Home,Gender,Exam_Score
0,23,84,Low,High,No,7,73,Low,Yes,0,Low,Medium,Public,Positive,3,No,High School,Near,Male,67
1,19,64,Low,Medium,No,8,59,Low,Yes,2,Medium,Medium,Public,Negative,4,No,College,Moderate,Female,61
2,24,98,Medium,Medium,Yes,7,91,Medium,Yes,2,Medium,Medium,Public,Neutral,4,No,Postgraduate,Near,Male,74
3,29,89,Low,Medium,Yes,8,98,Medium,Yes,1,Medium,Medium,Public,Negative,4,No,High School,Moderate,Male,71
4,19,92,Medium,Medium,Yes,6,65,Medium,Yes,3,Medium,High,Public,Neutral,4,No,College,Near,Female,70


## Review Distinct Values

Count the number of different values in each column.
The `Unknown` label is included as a distinct value wherever it appears.

In [5]:
# Count distinct values and sort from largest to smallest
df.nunique().sort_values(ascending=False).to_frame(name="Distinct Values")

,Distinct Values
Previous_Scores,51
Exam_Score,45
Attendance,41
Hours_Studied,41
Tutoring_Sessions,9
Sleep_Hours,7
Physical_Activity,7
Teacher_Quality,4
Parental_Education_Level,4
Distance_from_Home,4


### Findings

- Previous_Scores has the most distinct values (51), followed by
  Exam_Score (45).
- Attendance and Hours_Studied each have 41 distinct values.
- Teacher_Quality, Parental_Education_Level and Distance_from_Home
  each have four categories, including `Unknown` added during cleaning.
- Five columns have two categories each: Extracurricular_Activities,
  Internet_Access, School_Type, Learning_Disabilities and Gender.
- `Unknown` represents missing information, so it should not be
  interpreted as a measured level when comparing groups.

## Review Unknown Categories

Missing entries were labelled `Unknown` during cleaning.
Count these entries and calculate their percentage of the dataset
to inform later comparisons.

In [9]:
# Select the columns where missing values were labelled Unknown
# Based on reference materials from GeeksforGeeks and the pandas documentation
# (pandas.pydata.org), adapted for this project with ChatGPT assistance.
unknown_columns = [
    "Teacher_Quality",
    "Parental_Education_Level",
    "Distance_from_Home"
]

# Count Unknown entries in each column
unknown_counts = df[unknown_columns].eq("Unknown").sum()

# Display counts and percentages, largest first
unknown_summary = pd.DataFrame({
    "Unknown Count": unknown_counts,
    "Percentage (%)": (unknown_counts / len(df) * 100).round(2)
})

unknown_summary.sort_values("Unknown Count", ascending=False)

,Unknown Count,Percentage (%)
Parental_Education_Level,90,1.36
Teacher_Quality,78,1.18
Distance_from_Home,67,1.01


### Findings

- Parental_Education_Level has 90 Unknown entries (1.36%).
- Teacher_Quality has 78 Unknown entries (1.18%).
- Distance_from_Home has 67 Unknown entries (1.01%).
- Each column has fewer than 1.5% Unknown entries. These may overlap
  across the same students, so the counts cannot be added to calculate
  the number of affected students.
- All rows remain in the cleaned dataset. Unknown entries will be
  identified separately in categorical summaries and will not be
  treated as an ordered level such as Low, Medium or High.

## Descriptive Statistics

Summarise the numerical columns using counts, means, standard deviations,
minimums, quartiles and maximums. These statistics help describe typical
values and how widely the data varies.

In [10]:
# Summarise numerical columns, with one row per variable
df.describe().T.round(2)

,count,mean,std,min,25%,50%,75%,max
Hours_Studied,6607.0,19.98,5.99,1.0,16.0,20.0,24.0,44.0
Attendance,6607.0,79.98,11.55,60.0,70.0,80.0,90.0,100.0
Sleep_Hours,6607.0,7.03,1.47,4.0,6.0,7.0,8.0,10.0
Previous_Scores,6607.0,75.07,14.40,50.0,63.0,75.0,88.0,100.0
Tutoring_Sessions,6607.0,1.49,1.23,0.0,1.0,1.0,2.0,8.0
Physical_Activity,6607.0,2.97,1.03,0.0,2.0,3.0,4.0,6.0
Exam_Score,6607.0,67.24,3.89,55.0,65.0,67.0,69.0,101.0


### Findings

- All seven numerical columns contain 6,607 values, with no missing entries.
- Hours_Studied has a mean of 19.98 and a median of 20.
  The middle 50% of values lie between 16 and 24.
- Attendance averages 79.98%, with values ranging from 60% to 100%.
- Sleep_Hours has a mean of 7.03 and a median of 7.
- Previous_Scores averages 75.07, while Exam_Score averages 67.24.
  This difference alone does not establish a decline in performance,
  as the assessments may differ.
- Tutoring_Sessions has a median of 1 and ranges from 0 to 8.
- Physical_Activity has a mean of 2.97 and a median of 3.
- Exam_Score has a median of 67, with the middle 50% between 65 and 69.
  Its standard deviation is 3.89.
- The maximum Exam_Score of 101 remains flagged as an unresolved
  data quality issue and has been retained.